### PR TITLE
ci: add optimize unit tests to unified CI workflow

### PR DESCRIPTION
**Pull request by misiekhardcore**
## Description

After merging optimize to the monorepo, there are no Optimize related checks performed to ensure that merged PRs are not breaking Optimize. This has already happened with some renovate PRs that had automerged and broke up Optimize.
 Not the whole Optimize CI workflow can be used as a requred check, cause in the current state it takes too much time to perform, but running at least the unit tests will give us more security

## Checklist

- [x] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [x] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

relates to #21072

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (Zelldon) [inline .github/workflows/ci.yml:427]: ❌ Please take a look at the other jobs. We need to restrict the execution via the path filtering, otherwise it is always run.

Furthermore, right now it is not considered in the check results. That means even it is running it doesn't matter as the result it just ignored.

Example from java checks
>   java-checks:
>     if: needs.detect-changes.outputs.java-code-changes == 'true'
>     needs: [detect-changes]

**Review by RomanJRW — DISMISSED** [review]

File: .github/workflows/ci.yml
Comment (misiekhardcore) [reply .github/workflows/ci.yml:427]: there is a path filtering in that workflow itself so I thought it makes no sense to duplicate it. Does it make sense or should I still add another path filter nayway?

Comment (OmranAbazid) [reply .github/workflows/ci.yml:427]: I think it should follow the pattern described here:
https://github.com/camunda/camunda/wiki/CI-&-Automation#implementation

**Review by OmranAbazid — COMMENTED** [review]
See [comment](https://github.com/camunda/camunda/pull/21063/files#r1715408065)

File: .github/workflows/ci.yml
Comment (cmur2) [reply .github/workflows/ci.yml:427]: Yes please align with and follow the pattern that Omran linked to have a common approach everywhere

**Review by cmur2 — COMMENTED** [review]
Make sure to also follow [other inclusion criteria](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria) such as integration with Vault and emitting CI health metrics

File: .github/workflows/optimize-unit-tests.yml
Comment (cmur2) [inline .github/workflows/optimize-unit-tests.yml:57]: Please use the code from https://github.com/camunda/camunda/wiki/CI-&-Automation#metrics-collection which uses `build_status: ${{ job.status }}`

File: .github/workflows/optimize-unit-tests.yml
Comment (cmur2) [inline .github/workflows/optimize-unit-tests.yml:67]: Can this change detection logic here be refactored into the common place in `.github/actions/paths-filter` to avoid 2 layers of change detection where one would be enough?

**Review by OmranAbazid — APPROVED** [review]

**Comment by cmur2** [comment]
@misiekhardcore it could be more straightforward to implement this PR by starting with an unmodified `ci.yml` to have a clean slate and be unbiased by existing code. Then you can add the `steps` from `optimize-unit-tests.yml` into it (copying, not calling) while following [the pattern mentioned before](https://github.com/camunda/camunda/wiki/CI-&-Automation#implementation) to avoid things like having duplicate change detection.

(Maybe this hint comes too late but for future PRs it could be a better approach)

**Comment by cmur2** [comment]
Don't merge this PR without addressing my comments please 🙏🏼

File: .github/workflows/optimize-unit-tests.yml
Comment (misiekhardcore) [reply .github/workflows/optimize-unit-tests.yml:67]: You mean just moved from the `optimzie-paths-filter` to `paths-filter`?

**Comment by misiekhardcore** [comment]
> @misiekhardcore it could be more straightforward to implement this PR by starting with an unmodified `ci.yml` to have a clean slate and be unbiased by existing code. Then you can add the `steps` from `optimize-unit-tests.yml` into it (copying, not calling) while following [the pattern mentioned before](https://github.com/camunda/camunda/wiki/CI-&-Automation#implementation) to avoid things like having duplicate change detection.
> 
> (Maybe this hint comes too late but for future PRs it could be a better approach)
> 
> EDIT: The duplication of the change detection also means that neither Optimize FE nor BE tests are run on this PR which changes GHA code that affects them, because the 2nd layer of change detection doesn't check for GHA code changes.

I can copy over the content of the unit tests workflow, but this will cause code duplication. I pushed new commit with the `paths-filter` update and adding the job to the list of `needs` in `check-results`.

File: .github/workflows/optimize-unit-tests.yml
Comment (cmur2) [reply .github/workflows/optimize-unit-tests.yml:67]: I also meant to not execute `./.github/actions/paths-filter` twice (once in `ci.yml` and once in this file) since that is redundant and thus unexpected.

**Comment by cmur2** [comment]
> I can copy over the content of the unit tests workflow, but this will cause code duplication.

I assumed that `optimize-unit-tests.yml` can be deleted afterwards (and not called anymore from `optimize-ci.yml` since it is now called from `ci.yml` instead) thus no code duplication.

**Comment by misiekhardcore** [comment]
@cmur2 I moved the unit tests to `ci.yml` as suggested and removed the unit tests workflow completely. The `build-platform-frontend` fails for Optimize because of breaking renovate update. This will be fixed in #21130. When changes here will be approved, I'll merge that PR first, then rebase this one and see if the build is fixed and then we can merge this

**Review by cmur2 — APPROVED** [review]

**Comment by misiekhardcore** [comment]
@cmur2 any idea why is this failing in the merge queue? I see that some zeebe jobs are failing and optimize backend is failing because of timing out, but in this PR, the optimize backend job takes only like 3 minutes

**Comment by cmur2** [comment]
>  EDIT: I don't know how this works, but yesterday evening these BE unit tests with 4 threads took 3mins, and today they seem to run longer than 10

Just a theory/hypothesis: It might be related to caching that makes it (sometimes) run fast if e.g. a Maven cache exists. If that is true than we still need to make sure it also runs when there is no cache available to speed it up.

**Comment by misiekhardcore** [comment]
There is no build output when the operation gets canceled due to the time out. Its really hard to figure out what is the bottleneck here. The only thing I've noticed is that sometimes when it starts running test command, it hands for a while on the fist step: `Scanning for projects...`. But I can't find anything about why it takes a minute or two sometimes on this, and sometimes it just takes a second

**Comment by cmur2** [comment]
> java checks are failing on a file I didn't modify

Fixed in https://github.com/camunda/camunda/pull/21172

**Comment by cmur2** [comment]
> There is no build output when the operation gets canceled due to the time out.

There is for the steps before the timeout happens, no? I meant to compare that timing with the timings of the runs that you mentioned finished in under 3-4 minutes and figure out what the differences are.

Chances are when merging it like this that it will block many many people when it runs into timeouts again - we need to make sure to consistenly stay below timeouts before attempting merge.

**Comment by cmur2** [comment]
@misiekhardcore you can rebase on `main` now

**Comment by misiekhardcore** [comment]
@cmur2 I re-ran these backend unit tests couple of times and they seem to pass each time. Maybe we should give it a try?

**Comment by cmur2** [comment]
> @cmur2 I re-ran these backend unit tests couple of times and they seem to pass each time. Maybe we should give it a try?

Sorry for the late reply, was out sick. Sure lets attempt to merge once CI is green - currently failing Optimize FE unit tests (maybe due to dependency upgrade?)

**Comment by misiekhardcore** [comment]
@cmur2 yes, some breaking deps updates were made. I will open a PR to fix those first

**Comment by misiekhardcore** [comment]
@cmur2 @Zelldon zeebe smoke tests keep timing out, is this a known issue?

**Review by cmur2 — APPROVED** [review]

**Review by buccarel — APPROVED** [review]

**Comment by cmur2** [comment]
> @cmur2 @Zelldon zeebe smoke tests keep timing out, is this a known issue?

Not to me and it does not happen on `main` either (builds there take 5mins or less, [example](https://github.com/camunda/camunda/actions/runs/10482011366/job/29032432350)). These timeouts prevent this PR from passing the merge queue successfully and I'm not sure why they happen on this PR 🤔 

What I see in the logs for this PR and what seems to take much time (9mins) is Maven messages like `Artifact com.fasterxml.jackson:jackson-bom:pom:2.17.2 is present in the local repository, but cached from a remote repository ID that is unavailable in current build context, verifying that is downloadable`...

**Comment by misiekhardcore** [comment]
looks like caching issues to me, but I believe these are not related to my changes?

**Comment by misiekhardcore** [comment]
@cmur2 @Zelldon any idea why this is still failing in zeebe tests? This PR is blocked for more than a week now because of those issues

**Comment by cmur2** [comment]
> @cmur2 @Zelldon any idea why this is still failing in zeebe tests? This PR is blocked for more than a week now because of those issues

I'm considering to increase the timeouts for the affected Zeebe jobs since Zeebe devs also could identify another cause for the unexpected long runtimes other than caching behavior (something which is to be addressed in https://github.com/camunda/camunda/issues/18750 ),

**Comment by cmur2** [comment]
So close on the last attempt but then there was a network error with DockerHub.

**Comment by cmur2** [comment]
@misiekhardcore We can try again or you can increase timeouts of the smoke tests to 25mins to see if that helps

**Comment by Zelldon** [comment]
> @misiekhardcore We can try again or you can increase timeouts of the smoke tests to 25mins to see if that helps

@cmur2 caches seem to be full again ~120 gig. Looks like we experience also in the smoke tests issues due to this maybe? 

Other than that I feel like it might be also related to the nodes itself. For example in the IT tests we build in ~ 1 min. In the other smoke tests it takes ~12 min.
In the failing case it was longer than 19 min

**Comment by cmur2** [comment]
> caches seem to be full again ~120 gig. Looks like we experience also in the smoke tests issues due to this maybe?

Not sure since historically cache usage up to ~300 - 500 gig has been working okayish (we still need to solve this in https://github.com/camunda/camunda/issues/18750 though).

> related to the nodes itself

It is mainly the smoke tests on macOs and Windows that cause troubles due to sporadic timeouts.

**Comment by misiekhardcore** [comment]
@Zelldon I can increase these timeouts, but then, will we have to revert them after this PR merges?

**Comment by cmur2** [comment]
> will we have to revert them after this PR merges

First let's see if that helps, we have anyways be prepared to revert the PR in case something bad happens (maybe the failures in the merge queue indicate an underlying problem that neither of us sees). Then later we can decide about the timeouts.

**Comment by cmur2** [comment]
Interesting that this worked now and would've stayed even within the old timings.

**Comment by misiekhardcore** [comment]
yeah, I think it would merge by luck this time, but it was very close to timing out for macOS tests

**Pull request merged**
